virtio-devices/src/console.rs

```diff
@@ -332,6 +473,46 @@ impl EpollHelperHandler for ConsoleEpollHandler {
         }
         Ok(())
     }
+
+    fn handle_timeout(&mut self, helper: &mut EpollHelper) -> Result<(), EpollHelperError> {
+        if !self.endpoint.is_pty() {
+            return Ok(());
+        }
+
+        if self.file_event_registered {
+            // This very specific case happens when the console is connected
+            // to a PTY. We know EPOLLHUP is always present when there's nothing
+            // connected at the other end of the PTY. That's why getting no event
+            // means we can flush the output of the console through the PTY.
+            self.trigger_pty_flush()
+                .map_err(EpollHelperError::HandleTimeout)?;
+        }
+
+        self.register_file_event(helper)
+    }
+
+    fn event_list(
+        &mut self,
+        helper: &mut EpollHelper,
+        events: &[epoll::Event],
+    ) -> Result<(), EpollHelperError> {
+        if self.file_event_registered {
+            for event in events {
+                if event.data as u16 == FILE_EVENT && (event.events & libc::EPOLLHUP as u32) != 0 {
+                    return Ok(());
+                }
+            }
+
+            // This very specific case happens when the console is connected
+            // to a PTY. We know EPOLLHUP is always present when there's nothing
+            // connected at the other end of the PTY. That's why getting no event
+            // means we can flush the output of the console through the PTY.
+            self.trigger_pty_flush()
+                .map_err(EpollHelperError::HandleTimeout)?;
+        }
+
+        self.register_file_event(helper)
```

Comment:
I wonder why we need to register the file event here? I thought it would be enough to check if the pty is connected every 500ms (after the last processed epoll event) by using `handle_timeout()`

Reply:
The problem if we rely exclusively on `handle_timeout()` is that other events (INPUT_QUEUE_EVENT, OUTPUT_QUEUE_EVENT, ...) can be triggered (especially at boot when the driver provides a bunch of descriptors to the backend), and they directly impact the ability to detect a connection because the timeout will not be triggered unless there NO event at all.